libraries/common/random.effekt

```diff
@@ -0,0 +1,177 @@
+module random
+
+import stream
+import io/error
+
+/// Infinite pull stream of random bytes.
+effect random(): Byte
+
+// ---------------------
+// Sources of randomness
+
+/// A streaming source (push stream) of byte-level randomness
+/// based on Park and Miller's MINSTD with revised parameters.
+///
+/// Deterministic: needs a 32bit `seed` -- you can use `bench::timestamp`.
+def minstd(seed: Int): Unit / emit[Byte] = {
+  // Initialize state with seed, ensuring it's not zero
+  var state = if (seed == 0) 1 else seed
+
+  def nextInt(): Int = {
+    // Uses only at most 32-bit integers internally
+    // (Schrage's method: https://en.wikipedia.org/wiki/Lehmer_random_number_generator#Schrage's_method)
+    val a = 48271
+    val m = 2147483647
+
+    val q = m / a           // 44488
+    val r = m.mod(a)        //  3399
+
+    val div = state / q     // max: M / Q = A = 48271
+    val rem = state.mod(q)	// max: Q - 1     = 44487
+
+    val s = rem * a;        // max: 44487 * 48271 = 2147431977
+    val t = div * r;        // max: 48271 *  3399 =  164073129
+
+    val result = s - t
+    // keep the state positive
+    if (result < 0) result + m else result
+  }
+
+  while (true) {
+    state = nextInt()
+    val b = state.mod(256).toByte
+    do emit(b)
+  }
+}
+
+/// A thin wrapper over `minstd`, handling a reader of random bytes.
+///
+/// Deterministic: needs a 32bit `seed` -- you can use `bench::timestamp`.
+///
+/// Implementation is similar to `stream::source`, specialized for bytes and the `random` effect.
+def minstd(seed: Int) { randomnessReader: () => Unit / random }: Unit = {
+  var next = box { 255.toByte } // sentinel value
+  next = box {
+    try {
+      minstd(seed)
+      <> // safe: randomness generator cannot run out of numbers...
+    } with emit[Byte] { v =>
+      next = box { resume(()) }
+      v
+    }
+  }
+
+  try randomnessReader() with random {
+    resume(next())
+  }
+}
+
+/// CSPRNG from `/dev/urandom`, handling a reader of random bytes.
+/// Only works on Unix-like OSes!
+def devurandom { randomnessReader: () => Unit / random }: Unit / {Exception[IOError], stop} = {
```

Comment:
Is it intentional that this can raise `stop`? I would hope that `/dev/urandom` does not end 🤔

Reply:
(or we'd handle this somehow, e.g. reopening)

Reply:
I get `stop` from the `read[Byte]()`: I could also just throw a new IOError?

Reply:
Done, thanks